### PR TITLE
[Bug fix] roll: map COL_MAJOR shards to the right core for HEIGHT/WIDTH Sharding

### LINKED ISSUE #55683: roll: `shard_linear()` COL_MAJOR formula is only valid for BLOCK_SHARDED — OOB / wrong core on WIDTH/HEIGHT
Related but **not** this bug: [#55353](https://github.com/tenstorrent/tt-metal/issues/55353) (shift vs padded TILE extent).
Item 9  from https://github.com/tenstorrent/tt-metal/issues/51214

---

## Problem

Native sharded `ttnn.roll` maps a (row, col) cell to a core via `shard_linear()` in `roll_program_factory.cpp`. The **ROW_MAJOR** branch uses the shard-grid extent (`sr * n_shard_cols + sc`) and is correct for HEIGHT / WIDTH / BLOCK.

The **COL_MAJOR** branch uses `sc * grid_cols + sr`. That only matches how tt_metal enumerates cores when the layout is **BLOCK_SHARDED** (`n_shard_rows == grid_cols`). For other layouts it is the wrong index.

`validate_roll()` only rejects multi-range grids, so WIDTH/HEIGHT + COL_MAJOR still reach this factory. Existing COL_MAJOR tests use `grid_rows == 1`, where the two enumerations coincide, so CI stays green.

## How it fails

**WIDTH_SHARDED + COL_MAJOR**, one-row grid — host crash / OOB:

- Shape `[1, 1, 8, 16]`, shard `(8, 2)`, `CoreRange((0,0),(7,0))`
- `grid_cols=8`, `grid_rows=1`, `num_cores=8`, `sr==0`, `sc` runs `0..7`
- `shard_linear` returns `sc * 8` → `0, 8, …, 56`
- `all_transfers` has size 8 → write past the end
- `physical_cores[run_src_core / grid_cols][run_src_core % grid_cols]` indexes row `1..7` of a 1-row vector → heap corruption / crash

**HEIGHT_SHARDED + COL_MAJOR**, multi-row grid — silent wrong data:

- 16 shards on `CoreRange((0,0),(7,1))` (`grid_rows=2`)
- Formula yields `c = sr` (row-major core)
- tt_metal places shard `sr` on the `sr`-th core of the **column-major** enumeration
- Every shard is gathered from and written to the **wrong core**

## Location

`ttnn/cpp/ttnn/operations/data_movement/roll/device/roll_program_factory.cpp` (~line 191):

```cpp
auto shard_linear = [&](uint32_t row, uint32_t col) -> uint32_t {
    const uint32_t sr = row / shard_cells_h;
    const uint32_t sc = col / shard_cells_w;
    return row_major_orient ? sr * n_shard_cols + sc : sc * grid_cols + sr;
};
```

tt_metal maps shard `i` → `corerange_to_cores(grid, n, row_wise=(orientation==ROW_MAJOR))[i]` (`buffer_distribution_spec.cpp` / `buffer.cpp`): column-major `(gx = i / grid_rows, gy = i % grid_rows)` for COL_MAJOR.

## Suggested fix

Compute the shard linear index the same way for both orientations (`idx = sr * n_shard_cols + sc`), then convert to the core-enumeration index:

- COL_MAJOR: `gx = idx / grid_rows`, `gy = idx % grid_rows`, return `gy * grid_cols + gx` (reduces to the current formula for BLOCK)
- or `TT_FATAL` COL_MAJOR unless the layout is BLOCK_SHARDED

## Test plan

Add cases in `tests/ttnn/unit_tests/base_functionality/test_roll.py` (and/or `tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py`):

- WIDTH_SHARDED + COL_MAJOR, single-row grid (the OOB case)
- HEIGHT_SHARDED + COL_MAJOR on a grid with `grid_rows > 1` (wrong-core silent path)
- BLOCK_SHARDED + COL_MAJOR still matches `torch.roll` (no regression)
- Distinct fill so a wrong-core gather cannot pass by accident

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
Fixes #55683 (item 9 of #51214).

## Problem

`shard_linear()` returned `sc * grid_cols + sr` for COL_MAJOR, conflating the **buffer shard index** with the **core enumeration index** — two things that only coincide when the shard grid equals the core grid, i.e. BLOCK_SHARDED. HEIGHT/WIDTH + COL_MAJOR therefore read from the wrong core, or indexed past the end of the per-core tables. Detailed analysis in #55683.

## Fix

Split the two indices apart in `roll_program_factory.cpp`:

| Helper | Role |
|---|---|
| `shard_index(row, col)` (L200) | buffer shard index, `sr * n_shard_cols + sc` — orientation-independent |
| `shard_to_core(idx)` (L212) | shard index → core enumeration index, for `all_transfers` / `physical_cores` |
| `core_to_shard(c)` (L216) | inverse; gives the shard a core owns, for the DRAM bank/offset args (L495, L497) |

Plus a `TT_FATAL` (L189) for a shard count exceeding the core grid, which indexed out of bounds either way.

**Both helpers are the identity for ROW_MAJOR and reduce to the previous expression for BLOCK_SHARDED, so no already-correct configuration changes behaviour.** Verified by enumerating old vs. new over every grid factorization.

## Points for reviewers

1. **`shard_index` is orientation-independent** (L200). This is the load-bearing claim — check it against the `core_to_host_pages` WIDTH/BLOCK branch, which advances `j_offset` before `i_offset`.
2. **ROW_MAJOR and BLOCK_SHARDED come out unchanged** — `shard_to_core` is the identity for the first and reduces to the old expression for the second. That is the whole blast-radius argument.
3. **`core_to_shard` (L216) must be the exact inverse of `shard_to_core`** (L212), since the DRAM bank/offset args (L495, L497) are derived through it.

## Verification & tests

Reproduced on Wormhole B0 and green after: HEIGHT + COL_MAJOR on a 2D core grid read whole shards from the wrong core, WIDTH + COL_MAJOR segfaulted in `create_descriptor`, BLOCK and every ROW_MAJOR config were and stay correct. Four new tests in `test_universal_input_tm_roll.py`, each run untilized and tilized, compare bit-exactly against `torch.roll` over a distinct-per-element fill and assert via the graph calltrace that the native factory ran rather than a fallback; the wrong-core and segfault cases all fail on the parent commit. **128 passed** (101 nightly + 27 base functionality).

Full before/after matrix, per-test coverage and parent-commit behaviour: https://github.com/tenstorrent/tt-metal/issues/55683#issuecomment-5581438966 — plus two scoping corrections to the issue as filed: https://github.com/tenstorrent/tt-metal/issues/55683#issuecomment-5580134317

---

### Pipeline status

- [x] Sanity
- [x] (Runtime) Sanity Test
- [x] Nightly for DM Category
- [x] All model test
- [x] Single card
- [x] T3K
- [x] Galaxy
- [x] (Tier 1) Models Device Perf Tests

---

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/roll)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/roll)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/roll)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/roll)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/roll)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/roll)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/roll)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/roll/device/roll_program_factory.cpp [hunk 1]
```diff
@@ -271,7 +292,7 @@ RollPlan compute_roll_plan(
             }
             all_transfers[run_dst_core].push_back(RollTransferDesc{
                 .src_physical_core = physical_cores[run_src_core / grid_cols][run_src_core % grid_cols],
-                .src_dram_shard_idx = run_src_core,
+                .src_dram_shard_idx = core_to_shard(run_src_core),
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/roll/device/roll_program_factory.cpp]: `run_src_core` is already `shard_to_core(shard_index(src_row, p.src_col))`, so this round-trips back to the shard index the caller had in hand. Carrying it in the run state (a `run_src_shard` beside `run_src_core`) would store it directly instead of relying on the two lambdas being exact inverses.

File: ttnn/cpp/ttnn/operations/data_movement/roll/device/roll_program_factory.cpp [hunk 2]
```diff
@@ -436,14 +457,14 @@ RollPlan compute_roll_plan(
         }
         TT_ASSERT(
```
Comment (Aswinmcw) [inline ttnn/cpp/ttnn/operations/data_movement/roll/device/roll_program_factory.cpp]: Since you're touching this statement: a violation here isn't a crash — `src_to_slot` hands the reader a slot index >= 2 and it reads outside `src_base[2]`, i.e. silently wrong output. `TT_ASSERT` compiles out in Release, which is exactly where that would ship. `TT_FATAL` would be the safer form.

File: tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py [hunk 3]
```diff
@@ -79,6 +79,48 @@ def run_roll(device, torch_input, layout, mem_config, shifts, dims):
...
+    spec = ttnn.ShardSpec(
+        ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(grid_x - 1, grid_y - 1))}),
+        (sh, sw),
+        orientation,
+    )
+    mem_config = ttnn.MemoryConfig(tensor_layout, ttnn.BufferType.L1, spec)
+    ttnn_input = ttnn.from_torch(
+        torch_input, dtype=ttnn.bfloat16, layout=layout, device=device, memory_config=mem_config
+    )
+    got = ttnn.to_torch(ttnn.roll(ttnn_input, list(shifts), list(dims)).cpu())
```
Comment (Aswinmcw) [inline tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py]: Worth asserting the output really came back sharded (`memory_config().memory_layout`/`shard_spec`) before the value compare. `native_ok` in `roll.cpp` is shape-dependent, so if a future gate routes these configs through the interleaved DRAM round-trip the tests keep passing while no longer exercising the factory this PR fixes.

File: tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py [hunk 4]
```diff
@@ -79,6 +79,48 @@ def run_roll(device, torch_input, layout, mem_config, shifts, dims):
...
+
+    Every element gets a distinct value, so a gather that reads the right offset from the *wrong*
+    core cannot coincide with the expected result. Values stay inside bfloat16's exactly
+    representable integer range (0..256), which is what lets this assert equality instead of PCC.
+    """
+    compute_grid = device.compute_with_storage_grid_size()
+    if grid_x > compute_grid.x or grid_y > compute_grid.y:
+        pytest.skip(f"Device grid ({compute_grid.x}x{compute_grid.y}) too small for {grid_x}x{grid_y}")
+    numel = int(torch.tensor(shape).prod())
+    assert numel <= 256, f"{numel} elements would alias in bfloat16; keep the distinct fill exact"
```
Comment (Aswinmcw) [inline tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py]: This cap pins the helper to ROW_MAJOR — a single 32x32 tile is already 1024 elements — so the multi-row-grid COL_MAJOR map is only exercised untilized, even though the mapping code is layout-independent. The wrong-core guarantee only needs values distinct *between* shards, so a per-shard fill (or a wider exact dtype) would let a TILE case in here too.

**Review by Aswinmcw — APPROVED** [review]
Mapping checks out against `corerange_to_cores`/`core_to_host_pages`. Four non-blocking suggestions inline.

File: tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py [hunk 5]
```diff
@@ -79,6 +79,70 @@ def run_roll(device, torch_input, layout, mem_config, shifts, dims):
...
+
+    ttnn.graph.begin_graph_capture(ttnn.graph.RunMode.NORMAL)
+    try:
+        ttnn_output = ttnn.roll(ttnn_input, list(shifts), list(dims))
+    finally:
+        captured = ttnn.graph.end_graph_capture()
+    # The native path is a single RollDeviceOperation. Both sharded fallbacks add device ops here:
+    # the TILE one an untilize/tilize pair, the RM one an interleaved slice/concat round-trip.
+    device_ops = [n for n in ttnn.graph.extract_calltrace(captured) if n.endswith("DeviceOperation")]
+    assert device_ops == ["RollDeviceOperation"], (
```
Comment (Aswinmcw) [inline tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py]: `roll.cpp` issues one `roll_sharded` per non-zero adjusted shift, so this exact-list compare bakes in a single-dim restriction that isn't stated anywhere. `assert device_ops and set(device_ops) == {"RollDeviceOperation"}` keeps the fallback detection (both fallbacks contribute untilize/tilize or reshard/slice/concat nodes) while staying correct if a caller ever passes multi-dim shifts.

File: tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py [hunk 6]
```diff
@@ -79,6 +79,70 @@ def run_roll(device, torch_input, layout, mem_config, shifts, dims):
...
+    assert numel <= 2**24, f"{numel} elements would alias in float32; keep the distinct fill exact"
+    torch_input = torch.arange(numel, dtype=torch.float32).reshape(shape)
+    spec = ttnn.ShardSpec(
+        ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(grid_x - 1, grid_y - 1))}),
+        (sh, sw),
+        orientation,
+    )
+    mem_config = ttnn.MemoryConfig(tensor_layout, ttnn.BufferType.L1, spec)
+    ttnn_input = ttnn.from_torch(
+        torch_input, dtype=ttnn.float32, layout=layout, device=device, memory_config=mem_config
```
Comment (Aswinmcw) [inline tests/ttnn/nightly/unit_tests/operations/data_movement/test_universal_input_tm_roll.py]: Moving the whole helper to float32 drops bfloat16 from these configs — only the new tilized shape needed the wider type; `[1, 2, 16, 8]` and `[1, 1, 8, 16]` were already exact in bf16. Since element size feeds `cell_size` and the `row_pitch_bytes` alignment, a `dtype` parameter (bf16 for the small RM cases, fp32 for the tile one) would cover both pitches instead of just the 4-byte one.

**Review by Aswinmcw — APPROVED** [review]
All four points from the previous round are addressed, and the graph-capture guard is a stronger check than the one I asked for — the reasoning that the fallbacks reshard back to the input config, so `memory_config()` alone proves nothing, is correct. Three minor notes below; none block.

**Comment by Aswinmcw** [comment]
Description needs a refresh for this push: it still says three new tests / **112 passed**, which predates the tilized twin, the float32 fill and the graph-capture guard. Reviewer point 2 also describes `flush()` converting a core index back to a shard index — that round-trip is gone now.

**Review by Aswinmcw — APPROVED** [review]
Both notes from the last round are addressed and the tilized twins now cover WIDTH and BLOCK too. C++ is unchanged since the previous approval; I re-derived the new test geometry (WIDTH `[1,1,64,256]` shard `(64,32)` → 8 shards over each of the four grids, BLOCK `[1,1,64,128]` shard `(32,32)` → a 2x4 shard grid whose COL_MAJOR placement is exactly the old expression) and it holds. No new comments.

One stale number: the description says **116 passed (89 nightly)**, which is the `b071c6d` count. Parametrizing WIDTH and BLOCK over layout took the four new tests from 25 cases to 37, so it should now be 128 (101 nightly).

**Review by mateusznowakTT — APPROVED** [review]

**Review by amorrisonTT — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]
